### PR TITLE
TLUNIZIN-148 getting the unpublished Courses List

### PR DESCRIPTION
**Pull request by pushyamig**
https://jira.it.umich.edu/browse/TLUNIZIN-148

@dlhaines @lsloan Due by: 09/30/15

### CODE REVIEW

File: src/main/java/edu/umich/tl/ApiCallHandler.java [hunk 1]
```diff
@@ -20,44 +20,69 @@
...
+		this.canvasCall=canvasCall;
 	}
 	
 	public enum CanvasApiEnum{
-		TERM, UNPUBLISHED_COURSE_LIST;
+		TERM, UNPUBLISHED_COURSE_LIST, UNPUBLISHED_COURSE_LIST_PAGINATION_URL;
 	}
 
-	public HttpResponse getApiResponse(CanvasApiEnum canvasApi) {
+	public HttpResponse getApiResponse(CanvasApiEnum canvasApi, String canvasTermIdForSisTermId, String url) {
```
Comment (lsloan) [inline src/main/java/edu/umich/tl/ApiCallHandler.java]: `CanvasApiEnum` and `canvasApi` are simultaneously too wordy and ambiguous names.  I suggest naming the enum `RequestTypeEnum` and the argument `requestType`.

**Comment by lsloan** [comment]
Make use of [method chaining](https://en.wikipedia.org/wiki/Method_chaining#Java) in your new classes, like `Course.java`, `CoursesForDelete.java`, `Term.java`, etc.  All of the setter methods should end with `return this;`.  Then this code:

``` java
Term term =new Term();
term.setStartDate(dateFormat.parse(Utils.dateChopper(startDate)));
term.setEndDate(dateFormat.parse(Utils.dateChopper(endDate)));
term.setCanvasTermId(String.valueOf((Integer)eachTerm.get("id")));
term.setSisTermId((String)eachTerm.get("sis_term_id"));
term.setTermName((String)eachTerm.get("name"));
```

could be reduced to:

``` java
Term term = new Term()
    .setStartDate(dateFormat.parse(Utils.dateChopper(startDate)))
    .setEndDate(dateFormat.parse(Utils.dateChopper(endDate)))
    .setCanvasTermId(String.valueOf((Integer)eachTerm.get("id")))
    .setSisTermId((String)eachTerm.get("sis_term_id"))
    .setTermName((String)eachTerm.get("name"));
```

It's not a huge reduction, but it does make the code easier to read because it's easier to see where the setting up of `term` begins and ends.

Of course, this doesn't need to be limited to new classes.   It's easy to update existing classes to use this style, too.

File: src/main/java/edu/umich/tl/Course.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package edu.umich.tl;
+
+public class Course {
+	private int id;
+	private String courseName;
+	private String startDate;
```
Comment (lsloan) [inline src/main/java/edu/umich/tl/Course.java:6]: In `Term.java`, you use a date type for `startDate` and `endDate`.  Why use a string here?

File: src/main/java/edu/umich/tl/CourseDelete.java [hunk 3]
```diff
@@ -62,32 +93,200 @@ private static void checkForApiType(Properties properties) {
...
+		HttpEntity entity = httpResponse.getEntity();
+		List<HashMap<String, Object>> courseList=new ArrayList<HashMap<String, Object>>();
+		try {
+			String jsonResponseString = EntityUtils.toString(entity);
+			courseList = mapper.readValue(jsonResponseString,new TypeReference<List<Object>>(){});
+			for (HashMap<String, Object> course : courseList) {
+				Course aCourse=new Course();
+				aCourse.setId((Integer)course.get("id"));
+				aCourse.setCourseName((String)course.get("name"));
+				aCourse.setStartDate((String)course.get("start_at"));
```
Comment (lsloan) [inline src/main/java/edu/umich/tl/CourseDelete.java]: The similar line for setting the start date of terms uses a date type instead of a string.  If the same were done here, it would make the code safer, because the date strings would need to be validated in order to turn them into date objects.

File: src/main/java/edu/umich/tl/Utils.java [hunk 4]
```diff
@@ -64,6 +66,13 @@ public static InputStream getInputStreamFromURL(String fileURL) {
 
 		return in;
 	}
+	public static String dateChopper(String dateAsString) {
```
Comment (lsloan) [inline src/main/java/edu/umich/tl/Utils.java]: It appears that this is trying to parse a date and time in ISO 8601 format.  It may be worthwhile to use an existing parser that can handle that format.  For example, `javax.xml.bind.DatatypeConverter.parseDateTime(dateAsString).getTime()` to get a Date object.  If you really need to use a string, append `.toString()`.

Comment (lsloan) [inline src/main/java/edu/umich/tl/Utils.java]: Also, you might want to look into using [joda-time](http://www.joda.org/joda-time/).  Anthony has used it and likes it.

File: src/main/java/edu/umich/tl/Course.java (hunk 2 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/Course.java:6]: I like the Method chaining comment. Never heard before. Done

File: src/main/java/edu/umich/tl/CourseDelete.java (hunk 3 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/CourseDelete.java]: All i need to check whether the  start_date and end_date fields is not `null`.So i just left it as string than date format.

File: src/main/java/edu/umich/tl/Course.java (hunk 2 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/Course.java:6]: All i need to check whether the start_date and end_date fields is not null.So i just left it as string than date format.

File: src/main/java/edu/umich/tl/ApiCallHandler.java (hunk 1 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/ApiCallHandler.java]: Done

File: src/main/java/edu/umich/tl/Utils.java (hunk 4 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/Utils.java]: Ok. This is good information. I don't need that date info from the canvas as i will be getting that info from the Mpathways or others. This information is not used anymore So i will take it out

File: src/main/java/edu/umich/tl/ApiCallHandler.java [hunk 5]
```diff
@@ -20,44 +20,69 @@
...
+			// we still need to make the check if this call is to 'direct' or 'esb' canvas. We can get info from the 'CanvasCallEnum'
+		case UNPUBLISHED_COURSE_LIST_PAGINATION_URL:
+			if (isThisIsADirectCanvasCall()) {
+				httpResponse=apiDirectCanvas(url);
+			}else if (isThisAEsbCanvasCall()) {
+				httpResponse=apiESBCanvas(url);
+			}
 			break;
 		default:
 			break;
```
Comment (dlhaines) [inline src/main/java/edu/umich/tl/ApiCallHandler.java:53]: If this is unexpected log a warning that an unknown request (xxx) was encountered.

File: src/main/java/edu/umich/tl/ApiCallHandler.java [hunk 6]
```diff
@@ -20,44 +20,69 @@
...
+			httpResponse = apiCallToCanvas(urlSuffix);
 			break;
 		case UNPUBLISHED_COURSE_LIST:
+			urlSuffix=API_VERSION+"/accounts/1/courses?enrollment_term_id="+canvasTermIdForSisTermId+"&published=false&"+PER_PAGE;
+			httpResponse = apiCallToCanvas(urlSuffix);
+			break;
+			// we are having separate case for unpublished course list since pagination object in the response header has fully framed url and hence we can use it directly,
+			// we still need to make the check if this call is to 'direct' or 'esb' canvas. We can get info from the 'CanvasCallEnum'
+		case UNPUBLISHED_COURSE_LIST_PAGINATION_URL:
+			if (isThisIsADirectCanvasCall()) {
```
Comment (dlhaines) [inline src/main/java/edu/umich/tl/ApiCallHandler.java]: Why not send this call through apiCallToCanvas?

It looks like the 'case' statement really only sets the URL to send to apiCallToCanvas.  If the api call is always made then it could be written only once and put right after the case statement.

File: src/main/java/edu/umich/tl/ApiCallHandler.java [hunk 7]
```diff
@@ -20,44 +20,69 @@
...
+		String url = null;
+		HttpResponse httpResponse = null;
+		if (isThisIsADirectCanvasCall()) {
+			url = canvasURL + urlSuffix;
+			httpResponse = apiDirectCanvas(url);
+		} else if (isThisAEsbCanvasCall()) {
+			url = esbURL + urlSuffix;
+			httpResponse =apiESBCanvas(url);
+		}
+		M_log.debug("The Api call URL: " + url);
```
Comment (dlhaines) [inline src/main/java/edu/umich/tl/ApiCallHandler.java]: Would it be easy to add the HTTP status code from the response? That would provide a clue in the log if something is going wrong.

File: src/main/java/edu/umich/tl/CourseDelete.java [hunk 8]
```diff
@@ -22,23 +34,31 @@
...
+		readPropertiesFromFile(args);
 		ApiCallHandler apiHandler=new ApiCallHandler(canvasCall);
-		getPreviousTerms(apiHandler);
+		CoursesForDelete coursesForDelete=new CoursesForDelete();
+		getTerms(apiHandler,coursesForDelete);
+		String canvasTermIdForSisTermId = getCanvasTermIdForSisTermId(coursesForDelete);
+		if(canvasTermIdForSisTermId==null) {
+			M_log.error("Cannot find \"CanvasTermId\" for \"sisTermId = "+sisTermIdFromPropsFile+
+					"\" provided. Please correct the \"sis.term.id\" property provided from \"canvasCourseDelete.properties\" file.");
+			System.exit(0);
```
Comment (dlhaines) [inline src/main/java/edu/umich/tl/CourseDelete.java]: This is an error so the exit code should not be 0.

File: src/main/java/edu/umich/tl/CourseDelete.java [hunk 9]
```diff
@@ -62,32 +88,193 @@ private static void checkForApiType(Properties properties) {
...
+		}  catch (IOException e) {
+			M_log.error("IOException occured getUnpublishedCourseList( ):" ,e);
 		}
 	}
+	/*
+	 * This helper method pull out the error message that is sent in case of error. Currently code distinguishes error
+	 * differently from directCanvas vs esbCanvas. This may not be always the case but ESB has Special Throttling message
+	 * than canvas
+	 */
+	private static String apiCallErrorHandler(HttpResponse httpResponse, String text, ApiCallHandler apiHandler) {
```
Comment (dlhaines) [inline src/main/java/edu/umich/tl/CourseDelete.java]: Make sure that the response code (and any other reasonable information) from the response is logged.

File: src/main/java/edu/umich/tl/CourseDelete.java [hunk 10]
```diff
@@ -62,32 +88,193 @@ private static void checkForApiType(Properties properties) {
...
+			//sending null for canvasTermId is fine since we get the url framed from the pagination object from the response header. 
+			httpResponse=apiHandler.getApiResponse(RequestTypeEnum.UNPUBLISHED_COURSE_LIST_PAGINATION_URL,null,url);
+		}else {
+			httpResponse = apiHandler.getApiResponse(RequestTypeEnum.UNPUBLISHED_COURSE_LIST,canvasTermIdForSisTermId,url);
+		}
+		ObjectMapper mapper = new ObjectMapper();
+		int statusCode = httpResponse.getStatusLine().getStatusCode();
+		if(statusCode!=200) {
+			M_log.error(apiCallErrorHandler(httpResponse,"Unpublished_Courses",apiHandler));
+			System.exit(0);
```
Comment (dlhaines) [inline src/main/java/edu/umich/tl/CourseDelete.java]: Use non-zero exit code.

File: src/main/java/edu/umich/tl/CourseDelete.java [hunk 11]
```diff
@@ -62,32 +88,193 @@ private static void checkForApiType(Properties properties) {
...
+			}
+			String nextPageUrl = getNextPageUrl(httpResponse);
+			if(nextPageUrl!=null) {
+				getUnpublishedCourseList(canvasTermIdForSisTermId,coursesForDelete,apiHandler,nextPageUrl);
+			}
+		} catch (JsonParseException e1) {
+			M_log.error("JsonParseException occured getUnpublishedCourseList() : ",e1);
+		} catch(JsonMappingException e1) {
+			M_log.error("JsonMappingException occured getUnpublishedCourseList() : ",e1);
+		}  catch (IOException e) {
```
Comment (dlhaines) [inline src/main/java/edu/umich/tl/CourseDelete.java:188]: indentation.

File: src/main/java/edu/umich/tl/CourseDelete.java [hunk 12]
```diff
@@ -62,32 +88,193 @@ private static void checkForApiType(Properties properties) {
...
+			M_log.error("IOException occured getUnpublishedCourseList( ):" ,e);
 		}
 	}
+	/*
+	 * This helper method pull out the error message that is sent in case of error. Currently code distinguishes error
+	 * differently from directCanvas vs esbCanvas. This may not be always the case but ESB has Special Throttling message
```
Comment (dlhaines) [inline src/main/java/edu/umich/tl/CourseDelete.java:194]: Note that throttling is not really an error, just a constraint that the program could accommodate.  Depending on the application it may be ok to treat it as an error.

File: src/main/java/edu/umich/tl/CourseDelete.java [hunk 13]
```diff
@@ -62,32 +88,193 @@ private static void checkForApiType(Properties properties) {
...
+     * <https://umich.test.instructure.com/api/v1/courses/26164/enrollments?type=TeacherEnrollment&page=1&per_page=100>; rel="first",
+     * <https://umich.test.instructure.com/api/v1/courses/26164/enrollments?type=TeacherEnrollment&page=1&per_page=100>; rel="last"
+     */
+	protected static String getNextPageUrl(HttpResponse response) {
+		M_log.debug("getNextPageUrl(): Called");
+		String result = null;
+		if (!response.containsHeader("Link")) {
+			return result;
+		}
+			Header[] linkHeaders = response.getHeaders("Link");
```
Comment (dlhaines) [inline src/main/java/edu/umich/tl/CourseDelete.java]: Indentation.

File: src/main/java/edu/umich/tl/ApiCallHandler.java (hunk 7 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/ApiCallHandler.java]: Good Idea

File: src/main/java/edu/umich/tl/CourseDelete.java (hunk 9 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/CourseDelete.java]: Now logging `Status code`.  error response returned from canvas is also logged before

File: src/main/java/edu/umich/tl/CourseDelete.java (hunk 11 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/CourseDelete.java:188]: Done

File: src/main/java/edu/umich/tl/CourseDelete.java (hunk 13 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/CourseDelete.java]: Done

File: src/main/java/edu/umich/tl/CourseDelete.java (hunk 12 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/CourseDelete.java:194]: My idea of having separate methods pulling the error response for canvasDirect, canvasESB is still reasonable?

File: src/main/java/edu/umich/tl/ApiCallHandler.java (hunk 5 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/ApiCallHandler.java:53]: Done

File: src/main/java/edu/umich/tl/CourseDelete.java (hunk 12 above)
Comment (dlhaines) [inline src/main/java/edu/umich/tl/CourseDelete.java:194]: Yes it is.

File: src/main/java/edu/umich/tl/CourseDelete.java (hunk 8 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/CourseDelete.java]: Changed to exit code `1`

File: src/main/java/edu/umich/tl/ApiCallHandler.java (hunk 6 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/ApiCallHandler.java]: I did not understand this Suggestion. can we talk F2F

File: src/main/java/edu/umich/tl/CourseDelete.java (hunk 10 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/CourseDelete.java]: I used exit code 1.

File: src/main/java/edu/umich/tl/ApiCallHandler.java (hunk 6 above)
Comment (pushyamig) [inline src/main/java/edu/umich/tl/ApiCallHandler.java]: Done
